nav2_map_server/src/vo_server/vector_object_server.cpp

```diff
@@ -0,0 +1,582 @@
+{
+  RCLCPP_INFO(get_logger(), "Deactivating");
+
+  if (map_timer_) {
+    map_timer_->cancel();
+    map_timer_.reset();
+  }
+  process_map_ = false;
+
+  map_pub_->on_deactivate();
+
+  // Destroying bond connection
+  destroyBond();
+
+  return nav2_util::CallbackReturn::SUCCESS;
+}
+
+nav2_util::CallbackReturn
+VectorObjectServer::on_cleanup(const rclcpp_lifecycle::State & /*state*/)
+{
+  RCLCPP_INFO(get_logger(), "Cleaning up");
+
+  add_shapes_service_.reset();
+  get_shapes_service_.reset();
+  remove_shapes_service_.reset();
+
+  map_pub_.reset();
+  map_.reset();
+
+  tf_listener_.reset();
+  tf_buffer_.reset();
+
+  return nav2_util::CallbackReturn::SUCCESS;
+}
+
+nav2_util::CallbackReturn
+VectorObjectServer::on_shutdown(const rclcpp_lifecycle::State & /*state*/)
+{
+  RCLCPP_INFO(get_logger(), "Shutting down");
+
+  return nav2_util::CallbackReturn::SUCCESS;
+}
+
+bool VectorObjectServer::transformVectorObjects()
+{
+  for (auto shape : shapes_) {
+    if (shape->getFrameID() != frame_id_ && !shape->getFrameID().empty()) {
+      // Shape to be updated dynamically
+      if (!shape->toFrame(frame_id_, tf_buffer_, transform_tolerance_)) {
+        RCLCPP_ERROR(
+          get_logger(), "Can not transform vector object from %s to %s frame",
+          shape->getFrameID().c_str(), frame_id_.c_str());
+        return false;
+      }
+    }
+  }
+
+  return true;
+}
+
+void VectorObjectServer::getMapBoundaries(
+  double & min_x, double & min_y, double & max_x, double & max_y)
+{
+  min_x = std::numeric_limits<double>::max();
+  min_y = std::numeric_limits<double>::max();
+  max_x = std::numeric_limits<double>::lowest();
+  max_y = std::numeric_limits<double>::lowest();
+
+  double min_p_x, min_p_y, max_p_x, max_p_y;
+
+  for (auto shape : shapes_) {
+    shape->getBoundaries(min_p_x, min_p_y, max_p_x, max_p_y);
+    if (min_p_x < min_x) {
+      min_x = min_p_x;
+    }
+    if (min_p_y < min_y) {
+      min_y = min_p_y;
+    }
+    if (max_p_x > max_x) {
+      max_x = max_p_x;
+    }
+    if (max_p_y > max_y) {
+      max_y = max_p_y;
+    }
+  }
+
+  if (
+    min_x == std::numeric_limits<double>::max() ||
+    min_y == std::numeric_limits<double>::max() ||
+    max_x == std::numeric_limits<double>::lowest() ||
+    max_y == std::numeric_limits<double>::lowest())
+  {
+    throw std::runtime_error("Can not obtain map boundaries");
+  }
+}
+
+void VectorObjectServer::updateMap(
+  const double & min_x, const double & min_y, const double & max_x, const double & max_y)
+{
+  // Calculate size of update map
+  int size_x = static_cast<int>((max_x - min_x) / resolution_) + 1;
+  int size_y = static_cast<int>((max_y - min_y) / resolution_) + 1;
+
+  if (size_x < 0) {
+    throw std::runtime_error("Incorrect map x-size");
+  }
+  if (size_y < 0) {
+    throw std::runtime_error("Incorrect map y-size");
+  }
+
+  if (!map_) {
+    map_ = std::make_shared<nav_msgs::msg::OccupancyGrid>();
```

Comment:
Perhaps some of this difficulty is because you're working with `nav_msgs::msg::OccupancyGrid` instead of the `nav2_costmap_2d::Costmap2D`? I believe that has the constructor to take in an Occ Grid + can be converted back out to an occ grid easily. That would give you access to all the convert world to map coordinates and 2D raytracing functions for free without need to put them into nav2_utils since its being used here too

Reply:
This is tied to the costmap2D layer anyway, so I don't think its at all a problem to use costmap2d here too

Reply:
The idea here is to make vector object server as an OccGrid map server for flexibility, e.g. for having used not only in Costmaps2D, but also in mind to have an ability to apply vector objects to/over OccupancyGrid-s. If we will move this to Costmap2D, there is not sense to keep vector object server inside map server package. So, if we want to use costmaps only, it should be moved in there.

Another question, is that OccGrid-s are being published and read by Costmap as standard message, that is already compatible with Costmap Filters (just subscribe to the OccGrid masks, published by VO-server, so it seem to me OK). If we want to switch to costmaps, we need to have a VO-node instead, embedded inside all necessary layers, and Costmap Filters code modification is needed. Is this the way we want to have it (it seem that initially we've agreed to have a separate VO-server, instead of embedding node).

Anyway, this is not a problem to change it to be embedded VO-node inside Costmaps2D, but I think, I need to understand an intention to be on the same page.

Reply:
>  for having used not only in Costmaps2D, but also in mind to have an ability to apply vector objects to/over OccupancyGrid-s

Its still published out as an `OccupancyGrid` message, but you can take advantage that the Costmap2D object internally that has the same kind of data and alot of useful utilities that you can leverage (e.g. raytracing, world to map conversions, default values, etc) rather than abstracting that out or storing the same kinds of information as independent objects. 

>  If we will move this to Costmap2D, there is not sense to keep vector object server inside map server package.

I would agree. The map server location wasn't my first thought about where this would live. 

> If we want to switch to costmaps, we need to have a VO-node instead, embedded inside all necessary layers, and Costmap Filters code modification is needed. Is this the way we want to have it (it seem that initially we've agreed to have a separate VO-server, instead of embedding node)?

I think my remark above clarifies that. Use `nav2_costmap_2d::Costmap2D internal to the node in order to create the occupancy grid of the shapes. Then convert to the `OccupancyGrid` message type for publishing

Reply:
> Its still published out as an OccupancyGrid message, but you can take advantage that the Costmap2D object internally

OK, got it. However, this will cause the Nav2 to do unnecessary job twice. First time, we will convert costmaps -> to OccupancyGrid maps before publishing by VO-server. Then, if it will be used by costmaps, so it is being converted back to Costmaps by Keepout Filter (or Speed Filter or something else).

Moreover, Costmaps have no messages interfaces, while `OccupancyGrid.msg` is a far old-known standard, everyone works with it well. This is the same way, as we've used OccGrids for filter masks, to exist along with maps.

In this way, VO-server having the concept to be an independent module, working with vector objects and OccupancyGrids, and nothing common with Costmaps. All responsibilities for costmaps are taking by costmaps (Costmap Filter layers). If they will be re-worked, VO-server could be untouched - I see no problem with it.
 So I am still not sure, that using Costmaps for internal representation of what will be converted to Occupancy Grids - is a right idea or not?

> I would agree. The map server location wasn't my first thought about where this would live.

Initially, I've developing VO-server inside Costmap2D package. But after I've realized that there is nothing, that keeps it to work with Costmaps (please refer to the prev. sentence discourse), I've decided to move it to map_server package.
If we will decide to switch to Costmaps, there is no problem to move it back to `nav2_costmap_2d`

Reply:
I'm not saying don't publish to the costmap filter as a OccupancyGrid message, I'm asking you to use Costmap2D internally to this server to do operations so that you don't need to abstract out the map2world, world2map, getData, raycasting in 2D/3D and other functions into nav2_util which I do not want expanded outside of costmap2D given the road to come. 

The Costmap2D object contains all of these nice utilities which you'd like to use as a cost-map object. It seems purpose made for what you want to do. A good chunk of your parameters in the server are also natural constructor arguments for it 

```

  map_topic_ = getROSParameter(
    shared_from_this(), "map_topic", "vo_map").as_string();
  frame_id_ = getROSParameter(
    shared_from_this(), "frame_id", "map").as_string();
  resolution_ = getROSParameter(
    shared_from_this(), "resolution", 0.05).as_double();
  default_value_ = getROSParameter(
    shared_from_this(), "default_value", nav2_util::OCC_GRID_UNKNOWN).as_int();
  transform_tolerance_ = getROSParameter(
    shared_from_this(), "transform_tolerance", 0.1).as_double();
```

Reply:
Since last Nav2 WG, it was decided to keep VO-server to be based on OccupancyGrids to not make double Costmap2D->OccGrid->Costmap2D conversion and thus not loose performance and precision.
Also, it was decided to return back `getMaskData()` routine to `nav2_costmap_2d` as it used only there (and there is no sense to doubling the code), which was made in https://github.com/ros-planning/navigation2/pull/3930/commits/d6b9cf123c5bb1a26ab4ffbbed22356d402dcb85